src/blockfilter.cpp

```diff
@@ -0,0 +1,323 @@
+{
+    if (elements.size() >= (1ULL << 32)) {
+        throw std::invalid_argument("N must be <2^32");
+    }
+
+    m_N = elements.size();
+    m_F = m_N << m_P;
+
+    CVectorWriter stream(GCS_SER_TYPE, GCS_SER_VERSION, m_encoded, 0);
+
+    WriteCompactSize(stream, m_N);
+
+    if (elements.empty()) {
+        return;
+    }
+
+    BitStreamWriter<CVectorWriter> bitwriter(stream);
+
+    uint64_t last_value = 0;
+    for (uint64_t value : BuildHashedSet(elements)) {
+        uint64_t delta = value - last_value;
+        GolombRiceEncode(bitwriter, m_P, delta);
+        last_value = value;
+    }
+
+    bitwriter.Flush();
+}
+
+bool GCSFilter::Match(const GCSFilter::Element& element) const
+{
+    uint64_t query = HashToRange(element);
+
+    CVectorReader stream(GCS_SER_TYPE, GCS_SER_VERSION, m_encoded, 0);
+
+    // Seek forward by size of N
+    uint64_t N = ReadCompactSize(stream);
+    assert(N == m_N);
+
+    BitStreamReader<CVectorReader> bitreader(stream);
+
+    uint64_t value = 0;
+    for (uint64_t i = 0; i < m_N; i++) {
+        uint64_t delta = GolombRiceDecode(bitreader, m_P);
+        value += delta;
+
+        if (query == value) {
+            return true;
+        } else if (query < value) {
+            break;
+        }
+    }
+
+    return false;
+}
+
+bool GCSFilter::MatchAny(const std::set<Element>& elements) const
+{
+    const std::vector<uint64_t>&& queries = BuildHashedSet(elements);
+
+    CVectorReader stream(GCS_SER_TYPE, GCS_SER_VERSION, m_encoded, 0);
+
+    // Seek forward by size of N
+    uint64_t N = ReadCompactSize(stream);
+    assert(N == m_N);
+
+    BitStreamReader<CVectorReader> bitreader(stream);
+
+    uint64_t value = 0;
+    auto query_it = queries.begin();
+    for (uint64_t i = 0; i < m_N; i++) {
+        uint64_t delta = GolombRiceDecode(bitreader, m_P);
+        value += delta;
+
+        while (true) {
+            if (query_it == queries.end()) {
+                return false;
+            } else if (*query_it == value) {
+                return true;
+            } else if (*query_it > value) {
+                break;
+            }
+
+            query_it++;
+        }
+    }
+
+    return false;
+}
+
+static std::set<GCSFilter::Element> BasicFilterElements(const CBlock& block)
+{
+    std::set<GCSFilter::Element> elements;
+    for (const CTransactionRef& tx : block.vtx) {
+        // Include txid of each transaction.
+        const uint256& txid = tx->GetHash();
+        elements.emplace(txid.begin(), txid.end());
+
+        // Include previous outpoint of each input, except for coinbase.
+        if (!tx->IsCoinBase()) {
+            for (const CTxIn& txin : tx->vin) {
+                std::vector<unsigned char> ser_outpoint;
+                CVectorWriter(GCS_SER_TYPE, GCS_SER_VERSION, ser_outpoint, 0, txin.prevout);
+                elements.insert(std::move(ser_outpoint));
+            }
+        }
+
+        // Include all data pushes in output scripts.
```

Comment:
This does not seem up to date with the latest version of BIP158 (I'll review in full later).

Reply:
Fixed. Need to update the test vectors.